src/System.Text.Json/src/System/Text/Json/Serialization/CachedUtf8JsonWriter.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+
+namespace System.Text.Json.Serialization
+{
+    internal sealed class CachedUtf8JsonWriter
+    {
+        [ThreadStatic]
+        private static CachedUtf8JsonWriter s_cachedInstance;
+
+        private readonly Utf8JsonWriter _writer;
+
+#if DEBUG
+        private bool _inUse;
+#endif
+
+        private CachedUtf8JsonWriter(IBufferWriter<byte> stream, JsonWriterOptions writerOptions)
+        {
+            _writer = new Utf8JsonWriter(stream, writerOptions);
+        }
+
+        public static CachedUtf8JsonWriter Get(IBufferWriter<byte> stream, JsonWriterOptions writerOptions)
```

Comment:
Actually, it looks like this approach will not work since it doesn't honor different writer options pass in to the serializer (via `JsonSerializerOptions`, which includes `WriteIndented`).

We would need to create a new writer if the options didn't match the ones in the cached writer.

This approach worked within `Microsoft.AspNetCore.Http.Connections.Common` because they always had the same writer options.